pandas/core/dtypes/cast.py

```diff
@@ -359,22 +359,32 @@ def maybe_promote(dtype, fill_value=np.nan):
         if isinstance(fill_value, datetime) and fill_value.tzinfo is not None:
             # Trying to insert tzaware into tznaive, have to cast to object
             dtype = np.dtype(np.object_)
+        elif is_integer(fill_value) or (is_float(fill_value) and not isna(fill_value)):
+            dtype = np.dtype(np.object_)
```

Comment:
The non-docstring changes need to go in a different PR.

Reply:
same